File: creational/Python/Builder/builder.py
import abc
from itertools import product

class ProductA:
    list_content = []

    def print_all_values(self):
        for content in self.list_content:
            print(content)
    
    def add_part(self, part):
        self.list_content.append(part)



class AbstractBuilder:
    @abc.abstractmethod
    def add_part_a(self):
        pass

    @abc.abstractmethod
    def add_part_b(self):
        pass

    @abc.abstractmethod
    def add_part_c(self):
        pass
    
    @abc.abstractmethod
    def print_values(self):
        pass



class ConcreteBuilder(AbstractBuilder):
    product: ProductA

    def __init__(self) -> None:
        self.product = ProductA()
    
    def add_part_a(self):
        self.product.add_part("part_a")
    
    def add_part_b(self):
        self.product.add_part("part_b")

    def add_part_c(self):
        self.product.add_part("part_c")
    
    def print_values(self):
        self.product.print_all_values()



class Director:
    builder: AbstractBuilder

    def set_builder(self, m_builder):
        self.builder = m_builder

    def build_mvp(self):
        self.builder.add_part_a()
    
    def build_full_product(self):
        self.builder.add_part_a()
        self.builder.add_part_b()
        self.builder.add_part_c()




def client(director: Director):
    concrete_product: ConcreteBuilder = ConcreteBuilder()
    director.set_builder(concrete_product)

    director.build_full_product()
    concrete_product.print_values()


if __name__ == '__main__':
    director = Director()
    client(director)

File: creational/C++/Builder/Builder.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;
class AbstractBuilder {

    public:

    virtual void producePartA() = 0;
    virtual void producePartB() = 0;
    virtual void producePartC() = 0;
    virtual void printAllValues() = 0;
};

class Product {
    private:
        vector <string> s;
    public:
    
    Product() {

    }

    void addPart(string part) {
        s.__emplace_back(part);
        return ;
    }

    void printAllValues() {
        for(int i = 0 ; i< s.size() ; i ++) {
            cout << s[i] << endl;
        }
    }



};

class ConcreteBuilder : public AbstractBuilder {

    Product *p;

    public:

        ConcreteBuilder() {

            p = new Product();
        }
        void producePartA(){
            p->addPart("part_a");
        }
        void producePartB(){
           p->addPart("part_b"); 
        }

        void producePartC(){
            p->addPart("part_c"); 
        }

        void printAllValues() {
            p->printAllValues();
        }

};

class Director {

    AbstractBuilder *builder;

    public:

    Director() {

      
    }

    void setBuilder(AbstractBuilder *m) {
        builder = m;
    }

    void buildMVP(){
        builder->producePartA();
    }

    void buildFullProduct() {
        builder->producePartA();
        builder->producePartB();
        builder->producePartC();
    }
};



void client(Director &director) {

    AbstractBuilder *builder = new ConcreteBuilder();
    director.setBuilder(builder);

    director.buildFullProduct();
    builder->printAllValues();
    
}

int main() {
    Director d;
    client(d);
    return 0;
}